imports/wasi_snapshot_preview1/poll.go

```diff
@@ -65,95 +77,148 @@ func pollOneoffFn(ctx context.Context, mod api.Module, params []uint64) syscall.
 	}
 
 	// Eagerly write the number of events which will equal subscriptions unless
-	// there's a fault in parsing (not processing).
+	// there'e a fault in parsing (not processing).
 	if !mod.Memory().WriteUint32Le(resultNevents, nsubscriptions) {
 		return syscall.EFAULT
 	}
 
 	// Loop through all subscriptions and write their output.
 
+	var stdinSubs []stdinEvent
+	var timeout time.Duration = 1<<63 - 1 // max timeout
+	readySubs := 0
+
 	// Layout is subscription_u: Union
 	// https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#subscription_u
 	for i := uint32(0); i < nsubscriptions; i++ {
 		inOffset := i * 48
 		outOffset := i * 32
 
 		eventType := inBuf[inOffset+8] // +8 past userdata
-		var errno syscall.Errno        // errno for this specific event (1-byte)
+		// +8 past userdata +8 contents_offset
+		argBuf := inBuf[inOffset+8+8:]
+		userData := inBuf[inOffset : inOffset+8]
+
+		evt := &event{
+			eventType: eventType,
+			userData:  userData,
+			errno:     wasip1.ErrnoSuccess,
+			outOffset: outOffset,
+		}
+
 		switch eventType {
 		case wasip1.EventTypeClock: // handle later
-			// +8 past userdata +8 contents_offset
-			errno = processClockEvent(ctx, mod, inBuf[inOffset+8+8:])
+			newTimeout, err := processClockEvent(argBuf)
+			if err != 0 {
+				return err
+			}
+			if newTimeout < timeout {
+				timeout = newTimeout
+			}
+			writeEvent(outBuf, evt)
 		case wasip1.EventTypeFdRead, wasip1.EventTypeFdWrite:
-			// +8 past userdata +8 contents_offset
-			errno = processFDEvent(mod, eventType, inBuf[inOffset+8+8:])
+			fsc := mod.(*wasm.ModuleInstance).Sys.FS()
+			fd := le.Uint32(argBuf)
+
+			stdinReader := processFDEvent(fsc, fd, evt)
+			if stdinReader == nil {
+				// if stdinReader is not an interactive session, then write back immediately
+				readySubs++
+				writeEvent(outBuf, evt)
```

Comment:
It looks like `outBuf` isn't guaranteed to be zero-initialized? I think this should be initialized with zero at the beginning of this function.